### DISCUSSION TITLE
passing abi types to an object through a class

### DISCUSSION THREAD

**Discussion by glitch-txs** [wagmi-dev/abitype#140] [Q&A] [answered]
Hi, I'm needing a little bit of help, I'm trying to use abitype in a class that creates a Contract object that has a function called `write` where I'm passing the name of the contract function as a string and the params.

```ts
type CallInfo<funcName> = {
    name: funcName;
    params: any[];
}
```

 I'm trying to set abitype in the class constructor and trying to get the abi functions type in the `write` function argunments.

```ts

type CallInfo<funcName> = {
    name: funcName;
    params: any[];
}

export class InitContractTest {

  abi: unknown
  address: `0x${string}`
  chain: Chain
  contract: ethers.Contract

  constructor({address, abi, chain}: ContractInfo){
    this.abi = abi
    this.address = address
    this.chain = chain
    const rpcProvider = new ethers.providers.JsonRpcProvider(this.chain.rpcUrls[0])
    this.contract = new ethers.Contract(this.address, this.abi as ethers.ContractInterface, rpcProvider)
  }

  async write(callInfo: CallInfo<ExtractAbiFunctionNames<typeof this.abi, 'payable'>>, setStatus?: (status: string)=> void){
  
  //...call contract function
  }
```

I'm getting this error: `Type 'unknown' does not satisfy the constraint 'Abi'` when trying to extract the func names

**Comment by glitch-txs** [comment]
I got rid of the errors with this:

```ts
import { Abi, ExtractAbiFunctionNames } from 'abitype'

export type ContractInfo<ABI = any> = {
  address:`0x${string}`
  abi:ABI
  chain:Chain
}

export class InitContractTest<ABI extends Abi> {

  abi: ABI
  address: `0x${string}`
  chain: Chain
  contract: ethers.Contract

  constructor({address, abi, chain}: ContractInfo<ABI>){
    this.abi = abi
    this.address = address
    this.chain = chain
    const rpcProvider = new ethers.providers.JsonRpcProvider(this.chain.rpcUrls[0])
    this.contract = new ethers.Contract(this.address, this.abi as ethers.ContractInterface, rpcProvider)
  }

  async write(callInfo: CallInfo<ExtractAbiFunctionNames<typeof this.abi, 'payable'>>, setStatus?: (status: string)=> void){
  
  //...call contract function
  }
```

but I'm not getting autocomplete here:

```ts
import { abiERC20 } from './abiERC20'

  const newContract = new InitContractTest<typeof abiERC20>(contractInfo)

    const res = await newContract.read({
      name:'', // <-- no autocomplete here
      params:[]
    })
```

**Reply by Raiden1411** [reply]
Could you please share the abi you are using? Or create a minimal reproducible example?

Are you declaring the abi `as const`?

**Reply by Raiden1411** [reply]
You might also want to take a look at this

![image](https://github.com/wagmi-dev/abitype/assets/67233402/9cd396b3-aad5-43c9-8f29-1930d6562b3c)

**Reply by glitch-txs** [reply]
hi, thanks for helping me, here I created a minimal reproduction example, I'm passing the abi as const as well
https://github.com/glitch-txs/minimal-repro

**Reply by glitch-txs** [reply]
btw in the readme of the repo `ExtractAbiFunctions` is imported but `ExtractAbiFunctionNames` is used (and not imported)
![image](https://github.com/wagmi-dev/abitype/assets/66949816/e1cb3547-0f70-4038-9ce5-dc84c1c8e30a)

**Reply by Raiden1411** [reply, marked as answer]
The [abi](https://github.com/glitch-txs/minimal-repro/blob/main/abi.ts) that you are using is missing the `stateMutability` property and that's why you are not getting that autocomplete. The abi that you are using seems like it's from a old solidity version where `constants` were used to represent the mutability of a contract. That property is now considered deprecated and it was replaced by the `stateMutability`.

Abitype stills supports the constants properties but the `stateMutability` must always be present on the abi. You can use the provided `erc20Abi` from `abitype/test` for a generic erc20 abi. 

Made a pr on that repo to fix the problem.

**Reply by glitch-txs** [reply]
thanks a lot! <3
